### PR TITLE
Dev/finish activity serialization

### PR DESCRIPTION
**Pull request by tao558**
This PR adds protobuf serialization for the vital type "activity". Tests for the JSON and protobuf serialization are also added.

In order for the JSON tests to pass, the implementations of load and save for an activity were modified. Activity has two fields that are pointers, and in the previous implementation, it was assumed that they were always non-null. When one of these fields was null, a null pointer was dereferenced which caused a seg-fault.

Other small improvements were made as well:
1.) In a few unrelated tests, iterators were not being incremented.
2.) Renamed a misnamed unrelated test

### CODE REVIEW

**Comment by tao558** [comment]
Jenkins test this please

File: arrows/serialize/json/load_save.cxx [hunk 1]
```diff
@@ -624,10 +624,15 @@ void save( cereal::JSONOutputArchive& archive,
...
              ::cereal::make_nvp( "label", activity.label() ),
              ::cereal::make_nvp( "confidence", activity.confidence() ) );
-    save( archive, *activity.activity_type());
     save( archive, activity.start());
     save( archive, activity.end());
-    save( archive, *activity.participants() );
+
+    // These may be null
+    if ( activity.activity_type() )
+      save( archive, *activity.activity_type());
```
[the remark this answers is not in the document]
Comment (tao558) [reply arrows/serialize/json/load_save.cxx:633]: Fixed in 55651293d54bc6f866d9db56f45c41d631739c0d

File: arrows/serialize/json/load_save.cxx [hunk 2]
```diff
@@ -646,15 +655,54 @@ void load( cereal::JSONInputArchive& archive,
...
+    has_participants = false;
+  }
+
+  if ( has_act_type )
+  {
+    activity.set_activity_type( act_type );
+  }
+  else
+  {
+    activity.set_activity_type( nullptr );
```
Comment (as6520) [inline arrows/serialize/json/load_save.cxx:695]: Can we set activity type to nullptr [here](https://github.com/Kitware/kwiver/pull/1024/files#diff-09735ca8ba7ffc48e8004498c9f7d14bR677)?

File: arrows/serialize/json/load_save.cxx [hunk 3]
```diff
@@ -646,15 +655,54 @@ void load( cereal::JSONInputArchive& archive,
...
+  }
+
+  if ( has_participants )
+  {
+    activity.set_participants( participants );
+  }
+  else
+  {
+    activity.set_participants( nullptr );
```
Comment (as6520) [inline arrows/serialize/json/load_save.cxx:704]: Can we set participants to nullptr [here](https://github.com/Kitware/kwiver/pull/1024/files#diff-09735ca8ba7ffc48e8004498c9f7d14bR686)?

File: vital/types/protobuf/activity.proto [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+package kwiver.protobuf;
+
+import "class_map.proto";
+import "object_track_set.proto";
+import "timestamp.proto";
+
+message activity {
+  required int64 id = 1;
+  required string label = 2;
+  optional class_map classifications = 3;
```
Comment (as6520) [inline vital/types/protobuf/activity.proto:42]: Shouldn't this be called activity_type?

Comment (tao558) [reply vital/types/protobuf/activity.proto:42]: I don't believe so. This has the same fields as the activity vital type. activity_type was refactored out in favor of class_map, which already has protobuf bindings

File: arrows/serialize/json/load_save.cxx (hunk 2 above)
Comment (tao558) [reply arrows/serialize/json/load_save.cxx:695]: Yes, great idea.

File: vital/types/protobuf/activity.proto (hunk 4 above)
Comment (as6520) [reply vital/types/protobuf/activity.proto:42]: I was referring to `classifications`, should `classifications` be renamed to `activity_type`?

File: arrows/serialize/json/load_save.cxx (hunk 2 above)
Comment (tao558) [reply arrows/serialize/json/load_save.cxx]: Fixed in 560077cce47e58452637078afd1b6d92ed82c0ef

File: arrows/serialize/json/load_save.cxx (hunk 3 above)
Comment (tao558) [reply arrows/serialize/json/load_save.cxx]: Fixed in 560077cce47e58452637078afd1b6d92ed82c0ef

File: vital/types/protobuf/activity.proto (hunk 4 above)
Comment (tao558) [reply vital/types/protobuf/activity.proto:42]: In that case yeah it definitely could be. I'll change it.

Comment (tao558) [reply vital/types/protobuf/activity.proto]: Fixed in e6f9cce61663bf4ac8cd3aa68cd96ab59c8beeac

File: arrows/serialize/json/tests/test_serialize.cxx [hunk 5]
```diff
@@ -194,6 +339,8 @@ TEST( serialize, detected_object )
     {
       EXPECT_EQ( *(o_it->first), *(d_it->first) );
       EXPECT_EQ( o_it->second, d_it->second );
+      o_it++;
```
Comment (dstoup) [inline arrows/serialize/json/tests/test_serialize.cxx:342]: Considering these are tests and are less subject to performance requirements, you can disregard this comment with regards to this PR, but in general, it's best to pre-increment when possible. Post-increment requires an extra copy which can get costly when the iterator objects are large or when there are a lot of them.

File: arrows/serialize/json/tests/test_load_save.cxx [hunk 6]
```diff
@@ -71,6 +72,170 @@ int main(int argc, char** argv)
...
+  EXPECT_EQ( act.id(), act_dser.id() );
+  EXPECT_EQ( act.label(), act_dser.label() );
+  EXPECT_DOUBLE_EQ( act.confidence(), act_dser.confidence() );
+  EXPECT_EQ( act.start(), act_dser.start() );
+  EXPECT_EQ( act.end(), act_dser.end() );
+
+  // Check values in the retrieved class map
+  auto const act_type = act.activity_type();
+  auto const act_type_dser = act_dser.activity_type();
+  EXPECT_EQ( act_type->size(), act_type_dser->size() );
```
[the remark this answers is not in the document]
Comment (tao558) [reply arrows/serialize/json/tests/test_load_save.cxx:193]: I did this intentionally to make it easier to read those 3 lines. The 2nd arguments are all lined up for those expect_double_eq calls. If you would still prefer I change it, I am happy to do so.

File: arrows/serialize/json/load_save.cxx [hunk 7]
```diff
@@ -646,14 +655,35 @@ void load( cereal::JSONInputArchive& archive,
...
+
+  // Optional parameters aren't supported
+  // for JSON. activity_type and participants may or may not exist,
+  // so we check if an exception is thrown when we look for them.
+  // If an exception is thrown, we set those fields to null.
+  try
+  {
+    load( archive, *act_type );
+  }
+  catch( cereal::Exception& )
```
[the remark this answers is not in the document]
Comment (tao558) [reply arrows/serialize/json/load_save.cxx]: The latter. We're first checking whether those fields are set. If they're not, that signifies that they should be set to null. Fixed in c51b579f145749e25bac297306d50f8db025a625

File: arrows/serialize/json/tests/test_serialize.cxx (hunk 5 above)
Comment (tao558) [reply arrows/serialize/json/tests/test_serialize.cxx]: Good to know! This was easy enough to change in this case: a7e446f368e2708cefd14cdd367d795516350dda

**Comment by tao558** [comment]
Jenkins test this please

**Pull request merged**